include/boost/thread/detail/thread.hpp

```diff
@@ -539,17 +529,14 @@ namespace thread_detail {
 #if defined BOOST_THREAD_USES_DATETIME
         bool timed_join(const system_time& abs_time)
         {
-          struct timespec const ts=detail::to_timespec(abs_time);
-          return do_try_join_until(ts);
+          return do_try_join_until(detail::timespec_to_internal_clock(abs_time));
```

Comment:
I don't think we can convert system_time to internal_clock when internal_clock is steady_clock, as system_time is based on system_clock.

In this case, we should iterate as done in line 495

Reply:
I believe the timespec_to_internal_clock function handles this case. When the monotonic clock is enabled, it takes the difference between the abs_time and the current system time, and then adds that to the current steady clock, and that's what it returns. That ensures that we wait for the correct amount of time, even if the system clock jumps while we're waiting.

Reply:
Ok, I see for timespec_to_internal_clock. The name is not a good one. 

I don't think the user wants to wait a delay, otherwise it would had used a duration.
He want to join until a specific system_time. If the time is changed while joining, either the time would already be reached and then we need to return or the time is not reached yet, and then we need to iterate. 

The current implementation is not good neither.